wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/WsAgentWebSocketConnectionKeeper.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.wsagent.server;
+
+import static com.google.common.collect.Sets.newConcurrentHashSet;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_HEARTBEAT;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_PERIOD;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_SUBSCRIBE;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_UNSUBSCRIBE;
+
+import java.util.Set;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestHandlerConfigurator;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestTransmitter;
+import org.eclipse.che.commons.schedule.ScheduleRate;
+
+/**
+ * Sends requests to interested clients to confirm that connection isn't lost.
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionKeeper {
+  private final Set<String> endpointIds = newConcurrentHashSet();
+  private RequestTransmitter transmitter;
+  private RequestHandlerConfigurator configurator;
+
+  @Inject
+  public WsAgentWebSocketConnectionKeeper(
```

Comment:
Can you elaborate on what will happen if the browser will be closed forcefully? Will it be unsubscribed ever? With the set grow? How would `sendRequest` method work if several entries from endpointIds are browsers that are gone?

Reply:
This is a common issue, should be fixed by this https://github.com/eclipse/che/issues/6177.

Reply:
I've added cleaning of unresponsive IDs